src/lib/OS/SyscallHandler.cc

```diff
@@ -1,15 +1,712 @@
+
+  const SyscallInfo info = syscallQueue_.front();
+  ProcessStateChange stateChange;
+
+  switch (info.syscallId) {
+    case 29: {  // ioctl
+      int64_t fd = info.registerArguments[0].get<int64_t>();
+      uint64_t request = info.registerArguments[1].get<uint64_t>();
+      uint64_t argp = info.registerArguments[2].get<uint64_t>();
+
+      std::vector<char> out;
+      int64_t retval = ioctl(fd, request, out);
+
+      assert(out.size() < 256 && "large ioctl() output not implemented");
+      uint8_t outSize = static_cast<uint8_t>(out.size());
+      stateChange = {ChangeType::REPLACEMENT, {info.ret}, {retval}};
+      stateChange.memoryAddresses.push_back({argp, outSize});
+      stateChange.memoryAddressValues.push_back(
+          RegisterValue(reinterpret_cast<const char*>(out.data()), outSize));
+      break;
+    }
+    case 46: {  // ftruncate
+      uint64_t fd = info.registerArguments[0].get<uint64_t>();
+      uint64_t length = info.registerArguments[1].get<uint64_t>();
+      stateChange = {
+          ChangeType::REPLACEMENT, {info.ret}, {ftruncate(fd, length)}};
+      break;
+    }
+    case 48: {  // faccessat
+      int64_t dfd = info.registerArguments[0].get<int64_t>();
+      uint64_t filenamePtr = info.registerArguments[1].get<uint64_t>();
+      int64_t mode = info.registerArguments[2].get<int64_t>();
+      // flag component not used, although function definition includes it
+      int64_t flag = 0;
+
+      char* filename = new char[PATH_MAX_LEN];
+      return readStringThen(
+          filename, vAddrTranslation_(filenamePtr, info.processId),
+          PATH_MAX_LEN, [=](auto length) {
+            // Invoke the kernel
+            int64_t retval = faccessat(dfd, filename, mode, flag);
+            ProcessStateChange stateChange = {
+                ChangeType::REPLACEMENT, {info.ret}, {retval}};
+            delete[] filename;
+            concludeSyscall(stateChange);
+          });
+      break;
+    }
+    case 56: {  // openat
+      int64_t dirfd = info.registerArguments[0].get<int64_t>();
+      uint64_t pathnamePtr = info.registerArguments[1].get<uint64_t>();
+      int64_t flags = info.registerArguments[2].get<int64_t>();
+      uint16_t mode = info.registerArguments[3].get<uint16_t>();
+
+      char* pathname = new char[PATH_MAX_LEN];
+      return readStringThen(
+          pathname, vAddrTranslation_(pathnamePtr, info.processId),
+          PATH_MAX_LEN, [=](auto length) {
+            // Invoke the kernel
+            uint64_t retval = openat(dirfd, pathname, flags, mode);
+            ProcessStateChange stateChange = {
+                ChangeType::REPLACEMENT, {info.ret}, {retval}};
+            delete[] pathname;
+            concludeSyscall(stateChange);
+          });
+      break;
+    }
+    case 57: {  // close
+      int64_t fd = info.registerArguments[0].get<int64_t>();
+      stateChange = {ChangeType::REPLACEMENT, {info.ret}, {close(fd)}};
+      break;
+    }
```

Comment:
Should close not be thread specific (possibly in the actual `close` function)?

Reply:
This applies throughout syscalls that utilise a FileDescriptor

Reply:
Implicitly done, no changes needed